streaming/src/main/scala/org/apache/spark/streaming/util/BatchedWriteAheadLog.scala

```diff
@@ -0,0 +1,212 @@
+   */
+  override def write(byteBuffer: ByteBuffer, time: Long): WriteAheadLogRecordHandle = {
+    val promise = Promise[WriteAheadLogRecordHandle]()
+    walWriteQueue.offer(RecordBuffer(byteBuffer, time, promise))
+    try {
+      Await.result(promise.future.recover { case _ => null }(ThreadUtils.sameThread),
+        WAL_WRITE_STATUS_TIMEOUT.milliseconds)
+    } catch {
+      case e: TimeoutException =>
+        logWarning(s"Write to Write Ahead Log promise timed out after " +
+          s"$WAL_WRITE_STATUS_TIMEOUT millis for record.")
+        null
+    }
+  }
+
+  /**
+   * Read a segment from an existing Write Ahead Log. The data may be aggregated, and the user
+   * should de-aggregate using [[BatchedWriteAheadLog.deaggregate]]
+   *
+   * This method is handled by the parent WriteAheadLog.
+   */
+  override def read(segment: WriteAheadLogRecordHandle): ByteBuffer = {
+    parent.read(segment)
+  }
+
+  /**
+   * Read all the existing logs from the log directory.
+   *
+   * This method is handled by the parent WriteAheadLog.
+   */
+  override def readAll(): JIterator[ByteBuffer] = {
+    parent.readAll().asScala.flatMap(deaggregate).asJava
+  }
+
+  /**
+   * Delete the log files that are older than the threshold time.
+   *
+   * This method is handled by the parent WriteAheadLog.
+   */
+  override def clean(threshTime: Long, waitForCompletion: Boolean): Unit = {
+    parent.clean(threshTime, waitForCompletion)
+  }
+
+
+  /**
+   * Stop the batched writer thread, fulfill promises with failures and close parent writer.
+   */
+  override def close(): Unit = {
+    logInfo("BatchedWriteAheadLog shutting down.")
+    active = false
+    batchedWriterThread.interrupt()
+    batchedWriterThread.join()
+    fulfillPromises()
+    parent.close()
+  }
+
+  /**
+   * Respond to any promises that may have been left in the queue, to unblock receivers during
+   * shutdown.
+   */
+  private def fulfillPromises(): Unit = {
+    while (!walWriteQueue.isEmpty) {
+      val RecordBuffer(_, _, promise) = walWriteQueue.poll()
+      promise.success(null)
+    }
+  }
+
+  /** Start the actual log writer on a separate thread. Visible (protected) for testing. */
+  protected def startBatchedWriterThread(): Thread = {
+    val thread = new Thread(new Runnable {
+      override def run(): Unit = {
+        while (active) {
+          try {
+            flushRecords()
+          } catch {
+            case NonFatal(e) =>
+              logWarning("Encountered exception in Batched Writer Thread.", e)
+          }
+        }
+        logInfo("Batched WAL Writer thread exiting.")
+      }
+    }, "Batched WAL Writer")
+    thread.setDaemon(true)
+    thread.start()
+    thread
+  }
+
+  /** Write all the records in the buffer to the write ahead log. Visible for testing. */
+  protected def flushRecords(): Unit = {
+    try {
+      buffer.append(walWriteQueue.take())
+      val numBatched = walWriteQueue.drainTo(buffer.asJava) + 1
```

Comment:
Is the `take` being done to ensure that there is at least 1 record being written out? Isn't it better to just better to call `drainTo` and check if `numBatched > 0` ? That seems far more intuitive.